Add command to comment out a selection of text

christoph-heiss opened the pull request:
[Closes #100]

Hi!

This implements the `buffer::toggle_line_comment` for normal and select_line mode.
It either:
- Comments the selected lines out if any of the lines are not yet prefixed by the appriopriate comment indicator.
- Remove the comment prefix from any line applicable if all selected lines are already prefixed by the comment indicator.

I'm really used to that functionality to quickly comment in/out something, so I though I'd implement this. This works only for languages which have single-line comments. We could extend that to block comments, but I don't think that is much of a priority (at least for me).

I made [a small gif](https://giphy.com/gifs/ZiOh0JIcaDwtPjX5cl/html5) on how this functionality looks.

~~How a comment starts is defined in a Hashmap, based on the syntax scope of the buffer. I already added quite a few languages, I tried to pick the most likely-to-be-used. We can extend that list easily in the future as needed.
**Question:** Should there be a way for users to define their own entries in their preferences file? Could be useful I guess. 😅~~

Also, for now I assigned the command to <kbd>[</kbd>. I can of course happily change that if you want.

cpfiffer commented:
I like it!

jmacdonald commented on src/commands/buffer.rs:
```diff
@@ -468,6 +469,127 @@ pub fn outdent_line(app: &mut Application) -> Result {
+}
+
+fn add_line_comment(buffer: &mut Buffer, lines: &[(usize, &str)], offset: usize, comment_content: &str) {
+    for (line, _) in lines {
+        let target = Position { line: *line, offset };
```
```suggestion
    for (line_number, _) in lines {
        let target = Position { line: *line_number, offset };
```
I know it's at odds with the `Position` type's line/offset fields, but we're using `line` to represent _content_ in the function below, and it represents a number here, which is somewhat confusing. :slightly_smiling_face:

christoph-heiss replied:
Yeah, that's confusing, you're right. I'm gonna revise all of the naming 😅

jmacdonald commented on src/commands/buffer.rs:
```diff
@@ -468,6 +469,127 @@ pub fn outdent_line(app: &mut Application) -> Result {
+
+    // Get the range of lines we'll comment based on
+    // either the current selection or cursor line.
+    let lines = match app.mode {
```
Defining lines as a range here here, but then later re-defining it as a collection made the logic across both confusing. Can we call this `line_numbers` here for added clarity? :slightly_smiling_face:

christoph-heiss replied:
Same here, going to revise that.

jmacdonald commented on src/commands/buffer.rs:
```diff
@@ -468,6 +469,127 @@ pub fn outdent_line(app: &mut Application) -> Result {
+
+    let buffer_range_content = buffer.read(&buffer_range).ok_or(CURRENT_LINE_MISSING)?;
+
+    // Split the buffer content on newlines and filter out any zero-length lines.
```
While we _are_ using this logic to transform the line data portion of this, the `zip` is subtle and it's not entirely clear what we're producing at the end of this operation.

What do you think of something along the lines of "produce a collection of line numbers and content, but only for non-empty lines"?

christoph-heiss replied:
Sounds good, going to point out that operation explicitly.

jmacdonald reviewed (commented):
This is looking great! I just had a few questions as I was reviewing it, and some minor suggestions.

christoph-heiss commented on src/commands/buffer.rs:
```diff
@@ -468,6 +469,127 @@ pub fn outdent_line(app: &mut Application) -> Result {
+            let content = line.trim_start();
+            (content.starts_with(comment_content), line.len() - content.len())
+        })
+        .fold((true, usize::MAX), |result, (toggle, offset)| {
```
They all will be indented at the lowest common level. To demonstrate what I mean, I made a [short gif](https://giphy.com/gifs/ZiOh0JIcaDwtPjX5cl/html5). (Could have put that in the PR from the beginning, I guess.)
That's how I had the neovim plugin configured. I personally like that more than indenting each line at the start of its content, but that would be doable too. (Maybe even as a configuration item?)

The combination of reducing to the toggle switch and offset as tuple is only to avoid iterating all the lines twice. I could split that for clarity, but at cost.

jmacdonald replied:
Oh, of course; I don't know what I was thinking, that's exactly the behaviour we want. It seemed odd when reading the source code first, before having really thought about the behaviour.

christoph-heiss commented on src/commands/buffer.rs:
```diff
@@ -468,6 +469,127 @@ pub fn outdent_line(app: &mut Application) -> Result {
+        // you can use the command `buffer::display_current_scope`.
+        static ref COMMENT_MAP: HashMap<Scope, &'static str> = {
+            let mut m = HashMap::new();
+            m.insert(Scope::new("embedding.php").unwrap(), "// ");
+            m.insert(Scope::new("source.c").unwrap(), "// ");
+            m.insert(Scope::new("source.c++").unwrap(), "// ");
+            m.insert(Scope::new("source.cmake").unwrap(), "# ");
+            m.insert(Scope::new("source.go").unwrap(), "// ");
+            m.insert(Scope::new("source.js").unwrap(), "// ");
+            m.insert(Scope::new("source.lua").unwrap(), "-- ");
+            m.insert(Scope::new("source.makefile").unwrap(), "# ");
+            m.insert(Scope::new("source.python").unwrap(), "# ");
+            m.insert(Scope::new("source.rust").unwrap(), "// ");
+            m.insert(Scope::new("source.shell.bash").unwrap(), "# ");
+            m.insert(Scope::new("source.swift").unwrap(), "// ");
+            m.insert(Scope::new("source.toml").unwrap(), "# ");
+            m.insert(Scope::new("source.yaml").unwrap(), "# ");
+            m.insert(Scope::new("text.tex").unwrap(), "% ");
+            m.insert(Scope::new("text.tex.latex").unwrap(), "% ");
```
Yeah, at first I also looked into the syntax definitions if there was anything to infer them, but nothing really usable. The closest thing I found was a regex for comments, but that would mean parsing that etc. So I though using a stupid-simple list is probably the best thing.

Putting it into a configurable file is a good idea. Should that be then be part of the regular preferences file or a separate? That would then also solve my question about an user-configurable way for defining own entries.

jmacdonald replied:
I think the preferences file would be a good fit, as a `types.extension.line_comment_prefix` entry:

```yaml
types:
  rust:
    line_comment_prefix: //
```

We should assume a trailing space, and just add that as part of the prefix injection logic you've written. I can't think of a scenario where someone would want control of that and I'm all for sensible defaults.

As an aside, it looks like [syntect added support for parsing `.tmPreferences` files](https://github.com/trishume/syntect/pull/223). It's a shame that those bits of metadata don't live in the `.sublime-syntax` files themselves. We could eventually support both of those, but then we'd something to make importing those more intuitive; a problem for another time. :slightly_smiling_face:

christoph-heiss commented:
Just pushed a new spin of this PR, addressing all your suggestions.

This PR got definitely bigger than I expected!
Anyway, to properly support having the line comment prefix in the preferences, overridable by the user, I had to first move all our defaults from constant to a external yaml file. I've done that similiar to how `src/input/key_map/` does it, I hope that is fine.

But I also like this solution way more than having a hashmap in-code.

jmacdonald commented on src/commands/buffer.rs:
```diff
@@ -1670,4 +1766,188 @@ mod tests {
+    }
+
+    #[test]
+    fn toggle_line_comment_remove_multiple_with_unequal_indent_in_select_line_mode() {
```
This is really thorough; nicely done. :ok_hand:

jmacdonald commented on src/models/application/preferences/mod.rs:
```diff
@@ -15,17 +15,15 @@ const APP_INFO: AppInfo = AppInfo {
     author: "Jordan MacDonald",
 };
 const FILE_NAME: &str = "config.yml";
-const LINE_LENGTH_GUIDE_DEFAULT: usize = 80;
```
Migrating these and other defaults to the preferences file while we're at it is a really nice touch. :sparkles:

jmacdonald reviewed (approved):
This shaped up quite nicely. Fantastic test coverage and the improvements to how we handle default preferences in general is :heart_eyes:.

Thanks for the hard work, @christoph-heiss, and happy new year! :beers:
